<launch>
    <rosparam param="use_sim_time"> false </rosparam>

    <node pkg="tf" type="static_transform_publisher" name="gps_broadcaster_base_to_gps" args="-0.2 -0.355 0.18 0 0 0 1 base_link gps 100">
        <!-- <remap from="tf" to="tf_ekf"/> -->
    </node>
    <!-- x y z yaw pitch roll frame_id child_frame_id period_in_ms  -->

    <node pkg="rover" type="imu_to_enu.py" name="imu_to_enu" output="screen" />  
    <node pkg="rover" type="gps_to_pose.py" name="gps_to_pose" output="screen" />
    <node pkg="rover" type="ekf_odom_to_pose.py" name="ekf_odom_to_pose" output="screen" />
    <node pkg="rover" type="override_covariance.py" name="override_covariance" output="screen">
        <param name="imu_orien_cov_multiplier" value="10" />
        <param name="imu_ang_vel_cov_multiplier" value="10" />
        <param name="imu_lin_acc_cov_multiplier" value="10" />
        <param name="gps_cov_multiplier" value="1" />
        <param name="odom_pose_cov_multiplier" value="100000" />
        <param name="odom_twist_cov_multiplier" value="10000" />
    </node>

    <node pkg="imu_filter_madgwick" type="imu_filter_node" name="imu_filter_madgwick" output="screen" >
        <param name="publish_tf" value="false" />
        <param name="use_mag" value="true" />
        <param name="world_frame" value="enu" />
        <param name="orientation_stddev" value="0.2" />
        <remap from="imu/mag" to="/zed_node/imu/mag" />
        <remap from="imu/data_raw" to="/imu/enu" />
        <remap from="imu/data" to="/imu/orient" />
    </node>

    <node pkg="rover" type="heading_filter.py" name="heading_filter" output="screen">

        <param name="magnetic_declination_radians" value="-0.177764053"/>
        <!-- for toronto: -10.18513 => -0.177764053 (obtained jan 11, 2025) -->
        <param name="armlength" value="1.5" />

    </node>

    <node pkg="rover" type="rtabmap_init.py" name="rtabmap_init" output="screen" />
        
    <include file="$(find rtabmap_launch)/launch/rtabmap.launch">
        <arg name="rtabmap_args"  value="--delete_db_on_start"/>
        <!-- RGB-Depth -->
        <arg name="rgb_topic"  value="/zed_node/rgb/image_rect_color"/>
        <arg name="depth_topic"  value="/zed_node/depth/depth_registered"/>
        <arg name="camera_info_topic" value="/zed_node/rgb/camera_info"/>
        <arg name="wait_imu_to_init" value="true"/>
        <arg name="imu_topic"  value="/zed_node/imu/data"/>
        <arg name="frame_id" value="base_link"/>
        <arg name="odom_frame_id" value="odom"/>
        <arg name="map_frame_id" value="map"/>
        <arg name="approx_sync" value="false"/>
        <!-- Disable rtabmap_viz that's crashing -->
        <arg name="rtabmap_viz" value="false"/>
        <arg name="rviz" value="false"/>
        <!-- Disable transform publishing -->
        <arg name="publish_tf_map" value="false"/>
        <arg name="publish_tf_odom" value="false"/>
        <arg name="odom_tf_angular_variance" value="-1"/>
        <arg name="odom_tf_linear_variance" value="-1"/>

        <remap from="/tf" to="/tf_fake"/>

        <!-- Disable map optimization -->
        <param name="Reg/Strategy" type="string" value="0"/>
        <param name="Mem/STMSize" type="string" value="1"/>
        <param name="Mem/IncrementalMemory" type="string" value="false"/>
        <param name="RGBD/NeighborLinkRefining" type="string" value="false"/>
        <param name="Optimizer/Iterations" type="string" value="0"/>
        <param name="Mem/NotLinkedNodesKept" type="string" value="false"/>
    </include>

    <!-- Set rtabmap parameters after include -->
    <param name="/rtabmap/publish_tf" type="bool" value="false"/>
    <param name="/rtabmap_ros/publish_tf" type="bool" value="false"/>
    <param name="/rtabmap_node/publish_tf" type="bool" value="false"/>
    <param name="/rtabmap_viz/publish_tf" type="bool" value="false"/>
    <param name="/rtabmap/odom_frame_id" value="odom"/>
    <param name="/rtabmap/map_frame_id" value="map"/>
    <param name="/rtabmap/frame_id" value="base_link"/>
    
    <node pkg="robot_localization" type="ekf_localization_node" name="ekf_local_node" output="screen">
        <param name="publish_tf" value="true" />
        <param name="odom_frame" value="odom" />
        <param name="base_link_frame" value="base_link" />
        <param name="map_frame" value="map" />
        <param name="world_frame" value="odom" />
        <param name="frequency" value="15" />  
        <param name="sensor_timeout" value="0.2" />  
        <param name="two_d_mode" value="false" />
        <param name="print_diagnostics" value="true" />
        <remap from="odometry/filtered" to="/odom/ekf" />
        <param name="imu0" value="/imu/orient/override" />
        <param name="imu0_differential" value="false"/>
        <param name="imu0_relative" value="false"/>
        <param name="imu0_queue_size" value="10" />
        <param name="imu0_remove_gravitational_acceleration" value="true" />
        <rosparam param="imu0_config">[  false, false, false, 
                                        true, true, true,
                                        false, false, false, 
                                        true, true, true,
                                        true, true, true]</rosparam>
        <param name="odom0" value="/rtabmap/odom/override" />
        <param name="odom0_differential" value="false"/>
        <param name="odom0_relative" value="false"/>
        <param name="odom0_queue_size" value="10" />
        <rosparam param="odom0_config">[ true, true, true, 
                                        true, true, true,
                                        true, true, true,
                                        true, true, true,
                                        false, false, false]</rosparam>
        <param name="imu0_remove_gravitational_acceleration" value="true" />

    </node>

    <node pkg="robot_localization" type="ekf_localization_node" name="ekf_global_node" output="screen">
        <param name="publish_tf" value="true" />
        <param name="odom_frame" value="odom" />
        <param name="base_link_frame" value="base_link" />
        <param name="map_frame" value="map" />
        <param name="world_frame" value="map" />
        <param name="frequency" value="5" />  
        <param name="sensor_timeout" value="2" />  
        <param name="two_d_mode" value="false" />
        <param name="print_diagnostics" value="true" />
        <remap from="odometry/filtered" to="/map/ekf" />
        <param name="imu0" value="/imu/orient/override" />
        <param name="imu0_differential" value="false"/>
        <param name="imu0_relative" value="false"/>
        <param name="imu0_queue_size" value="10" />
        <param name="imu0_remove_gravitational_acceleration" value="true" />
        <rosparam param="imu0_config">[  false, false, false, 
                                        true, true, true,
                                        false, false, false, 
                                        true, true, true,
                                        true, true, true]</rosparam>
        <param name="odom0" value="/rtabmap/odom/override" />
        <param name="odom0_differential" value="false"/>
        <param name="odom0_relative" value="false"/>
        <param name="odom0_queue_size" value="10" />
        <rosparam param="odom0_config">[ true, true, true, 
                                        true, true, true,
                                        true, true, true,
                                        true, true, true,
                                        false, false, false]</rosparam>
        <param name="odom1" value="/navsat/gps" />
        <param name="odom1_differential" value="false"/>
        <param name="odom1_relative" value="false"/>
        <param name="odom1_queue_size" value="10" />
        <rosparam param="odom1_config">[ true, true, true, 
                                        true, true, true,
                                        true, true, true,
                                        true, true, true,
                                        false, false, false]</rosparam>

    </node>

    <node pkg="robot_localization" type="navsat_transform_node" name="navsat_transform_node" respawn="true">

        <param name="frequency" value="5" />

        <param name="magnetic_declination_radians" value="-0.177764053" />
        <!-- for toronto: -10.18513 => -0.177764053 (obtained jan 11, 2025) -->

        <param name="yaw_offset" value="0"/>
        <param name="wait_for_datum" value="true" />
        <rosparam param="datum">[43.660517, -79.396553, 0.0, map, base_link]</rosparam> 
        <!-- [lat, lon, heading, world frame, base link] -->

        <remap from="/imu/data" to="/fused_heading" />
        <remap from="/gps/fix" to="/calian_gnss/gps/override" />
        <remap from="/odometry/filtered" to="/map/ekf" />
        <remap from="/odometry/gps" to="/navsat/gps" />

    </node>

</launch>